File: config-modifier.py
import argparse
import os
import yaml

# Argument parsing setup
arg_parser = argparse.ArgumentParser()
arg_parser.add_argument('-k', '--key')
arg_parser.add_argument('-v', '--value')
parsed_args = arg_parser.parse_args()
input_key = parsed_args.key
input_value = parsed_args.value

# Get the current working directory
current_directory = os.getcwd()
yaml_file_paths = []

# Walk through the directory to find YAML files
for dirpath, dirnames, filenames in os.walk(current_directory):
    for filename in filenames:
        if filename.endswith('.yaml') or filename.endswith('.yml'):
            yaml_file_paths.append(os.path.join(dirpath, filename))

# Dictionary to store the contents of the YAML files
yaml_data = {}

# Load the contents of the YAML files
for filepath in yaml_file_paths:
    with open(filepath, 'r') as file:
        try:
            data = yaml.safe_load(file)
            yaml_data[filepath] = data
        except:
            print('Error loading YAML file')

# Function to replace the key-value pair in the YAML content
def update_key_value(filepath, data, key, value):
    if key is None or value is None:
        print('Please provide both key and value to update')
        return

    updated_data = modify_value(data, key, value)
    with open(filepath, 'w') as file:
        yaml.safe_dump(updated_data, file)

# Recursive function to update the key-value pair
def modify_value(data_structure, target_key, new_value):
    if isinstance(data_structure, dict):
        for k, v in data_structure.items():
            if k == target_key:
                data_structure[k] = new_value
            elif isinstance(v, (dict, list)):
                modify_value(v, target_key, new_value)
    elif isinstance(data_structure, list):
        for item in data_structure:
            modify_value(item, target_key, new_value)
    return data_structure

# Try to update the key-value pair in each YAML file
try:
    for filepath, data in yaml_data.items():
        update_key_value(filepath, data, input_key, input_value)
        print(f'Successfully updated key `{input_key}` to value `{input_value}` in {filepath}')
except:
    print('Failed to update files')

File: config-modifier.go
package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"gopkg.in/yaml.v3"
)

func main() {
	// Parse command line arguments
	key := flag.String("k", "", "Key to update")
	value := flag.String("v", "", "New value for the key")
	flag.Parse()

	if *key == "" || *value == "" {
		fmt.Println("Please provide both key and value to update")
		return
	}

	// Get the current working directory
	currentDirectory, err := os.Getwd()
	if err != nil {
		fmt.Println("Error getting current directory:", err)
		return
	}

	// Find all YAML files in the directory
	yamlFilePaths, err := findYAMLFiles(currentDirectory)
	if err != nil {
		fmt.Println("Error finding YAML files:", err)
		return
	}

	// Update each YAML file
	for _, filepath := range yamlFilePaths {
		if err := updateYAMLFile(filepath, *key, *value); err != nil {
			fmt.Println("Failed to update file:", filepath, err)
		} else {
			fmt.Printf("Successfully updated key `%s` to value `%s` in %s\n", *key, *value, filepath)
		}
	}
}

// Recursively find all YAML files in the given directory
func findYAMLFiles(dir string) ([]string, error) {
	var yamlFiles []string
	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if !info.IsDir() && (filepath.Ext(path) == ".yaml" || filepath.Ext(path) == ".yml") {
			yamlFiles = append(yamlFiles, path)
		}
		return nil
	})
	return yamlFiles, err
}

// Update the key-value pair in the given YAML file
func updateYAMLFile(filepath string, key, value string) error {
	data, err := ioutil.ReadFile(filepath)
	if err != nil {
		return err
	}

	var yamlNode yaml.Node
	if err := yaml.Unmarshal(data, &yamlNode); err != nil {
		return err
	}

	updateKeyValue(&yamlNode, key, value)

	updatedData, err := yaml.Marshal(&yamlNode)
	if err != nil {
		return err
	}

	return ioutil.WriteFile(filepath, updatedData, 0644)
}

// Recursively update the key-value pair in the YAML node
func updateKeyValue(node *yaml.Node, targetKey, newValue string) {
	if node.Kind == yaml.DocumentNode {
		for _, child := range node.Content {
			updateKeyValue(child, targetKey, newValue)
		}
	}
	if node.Kind == yaml.MappingNode {
		for i := 0; i < len(node.Content); i += 2 {
			k := node.Content[i]
			v := node.Content[i+1]
			if k.Value == targetKey {
				// Update the value while preserving its type
				if v.Tag == "!!int" {
					v.Value = newValue
					v.Tag = "!!int"
				} else if v.Tag == "!!bool" {
					v.Value = newValue
					v.Tag = "!!bool"
				} else if v.Tag == "!!float" {
					v.Value = newValue
					v.Tag = "!!float"
				} else {
					v.Value = newValue
					v.Tag = "!!str"
				}
				return
			}
			updateKeyValue(v, targetKey, newValue)
		}
	}
	if node.Kind == yaml.SequenceNode {
		for _, child := range node.Content {
			updateKeyValue(child, targetKey, newValue)
		}
	}
}

